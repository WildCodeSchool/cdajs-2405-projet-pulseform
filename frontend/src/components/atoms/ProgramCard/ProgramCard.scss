@use "@assets/variables" as *;

.program-card {
  all:unset; // remove all button native styles
  position: relative;
  overflow: hidden;
  border-radius: 10px;
  width: 23.25rem;
  max-width: 100%;
  height: 13.5rem;
  cursor: pointer;

  @media screen and (max-width: 768px) {
    width: 100%;
  }

  &__image {
    width: 100%;
    height: 100%;
    object-fit: cover;
    transition: transform 200ms ease;

    &:hover {
      transform: scale(1.1); /* Scale the image 110% on hover */
    }
  }

  &__overlay {
    display: flex;
    justify-content: space-between;
    flex-direction: column;
    padding: 1.5rem;
    inset: 0;
    z-index: 1;
    background: linear-gradient(rgba(0, 0, 0, 0.4), rgba(0, 0, 0, 0.4));
    position: absolute;
    pointer-events: none; /* Prevent overlay from blocking hover interactions on image */

    .program-card__header {
      display: flex;
      justify-content: flex-end;
      align-items: flex-end;
      flex-direction: column;
      gap: 0.15rem;
      color: $white;
    }

    .program-card__title_container {
      display: flex;
      flex-direction: row;
      justify-content: flex-start;

      .program-card__title {
        @include h2-phone;
        color: $white;
      }
    }
  }
}
