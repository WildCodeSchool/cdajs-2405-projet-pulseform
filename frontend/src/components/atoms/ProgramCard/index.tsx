import { forwardRef } from "react";
import { useNavigate } from "react-router-dom";

import DurationLabel from "../DurationLabel";
import FlamesFitnessLevel from "../FlamesFitnessLevelLabel";

import type { Program } from "@graphql/__generated__/schema";
import { convertSecondsToMinRounded } from "@utils/timeUtils";

import "./ProgramCard.scss";

type ProgramCardProps = {
  program: Program;
};

const ProgramCard = forwardRef<HTMLButtonElement, ProgramCardProps>(
  ({ program }, ref) => {
    const convertedDuration = convertSecondsToMinRounded(
      program.total_duration,
    );
    const navigate = useNavigate();

    return (
      <button
        type="button"
        className="program-card"
        onClick={() => navigate(`/program/${program.id}`)}
        ref={ref}
      >
        <img
          src={program?.image || ""}
          alt={program?.name || ""}
          className="program-card__image"
        />
        <div className="program-card__overlay">
          <div className="program-card__header">
            <DurationLabel duration={convertedDuration || ""} />
            <FlamesFitnessLevel level={program?.level} />
          </div>
          <div className="program-card__title_container">
            <h3 className="program-card__title">{program?.name}</h3>
          </div>
        </div>
      </button>
    );
  },
);

ProgramCard.displayName = "ProgramCard";

export default ProgramCard;
