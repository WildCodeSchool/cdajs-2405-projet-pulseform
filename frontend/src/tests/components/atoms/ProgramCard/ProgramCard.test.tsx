import ProgramCard from "@components/atoms/ProgramCard";
import { FitnessLevel, type Program } from "@graphql/__generated__/schema";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { beforeAll, describe, expect, it, vi } from "vitest";

describe("ProgramCard Component", () => {
  beforeAll(() => {
    vi.stubEnv("VITE_URL_BACK", "http://localhost:4000");
    vi.stubEnv("VITE_IMAGE_PROGRAMME", "programs");
  });

  const mockProgram: Program = {
    id: "1",
    name: "Test Program",
    total_duration: 60,
    level: FitnessLevel.Beginner,
    created_at: new Date(),
    visibility: 0,
    image: "test.jpg",
  };

  it("matches the snapshot", () => {
    const tree = render(
      <MemoryRouter>
        <ProgramCard program={mockProgram} />
      </MemoryRouter>,
    );
    expect(tree).toMatchSnapshot();
  });
});
