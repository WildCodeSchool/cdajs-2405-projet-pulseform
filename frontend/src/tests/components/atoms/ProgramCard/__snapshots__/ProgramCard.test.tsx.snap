// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgramCard Component > matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="program-card"
        type="button"
      >
        <img
          alt="Test Program"
          class="program-card__image"
          src="http://localhost:4000/programs/test.jpg"
        />
        <div
          class="program-card__overlay"
        >
          <div
            class="program-card__header"
          >
            <div
              class="duration-label"
            >
              <span />
              <span>
                1 min
              </span>
            </div>
            <div
              class="flames-fitness-level-container"
            >
              <div
                class="flames-list"
              >
                <span />
              </div>
            </div>
          </div>
          <div
            class="program-card__title_container"
          >
            <h3
              class="program-card__title"
            >
              Test Program
            </h3>
          </div>
        </div>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="program-card"
      type="button"
    >
      <img
        alt="Test Program"
        class="program-card__image"
        src="http://localhost:4000/programs/test.jpg"
      />
      <div
        class="program-card__overlay"
      >
        <div
          class="program-card__header"
        >
          <div
            class="duration-label"
          >
            <span />
            <span>
              1 min
            </span>
          </div>
          <div
            class="flames-fitness-level-container"
          >
            <div
              class="flames-list"
            >
              <span />
            </div>
          </div>
        </div>
        <div
          class="program-card__title_container"
        >
          <h3
            class="program-card__title"
          >
            Test Program
          </h3>
        </div>
      </div>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
